### PR TITLE
Task Page for Verifying Signed Images

### PR DESCRIPTION
**Pull request by PushkarJ**
Supersedes https://github.com/kubernetes/website/pull/31611
In support of https://github.com/kubernetes/release/issues/2383
Relevant KEP: https://github.com/kubernetes/enhancements/issues/3031
/sig security release docs
Notes:
- This PR adds some automation to generate all the signed image names
- We may have to update this as we add more images that are signed

Suggestion for a better placement of this source of truth of signed images is welcome!

/cc @saschagrunert @puerco @sftim

### CODE REVIEW

**Comment by PushkarJ** [comment]
Fixes https://github.com/kubernetes/website/issues/31420
(forgot this earlier)

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (jihoon-seo) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:18]: Preview: https://deploy-preview-32184--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/verify-signed-images/

```suggestion
For a complete list of images that are signed please refer to [releases](/releases/download/) page.
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (jihoon-seo) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:27]: nit: A period is missing.

```suggestion
To verify all signed control plane images, please run this command.
```

File: content/en/releases/download.md
Comment (jihoon-seo) [inline content/en/releases/download.md:75]: Preview: https://deploy-preview-32184--kubernetes-io-vnext-staging.netlify.app/releases/download/

nit: A period is missing.

```suggestion
Full list of images that are signed can be found [here](/examples/admin/signed-images/auto-generated-list-of-all-signed-images.txt).
```

**Review by jihoon-seo — COMMENTED** [review]
Some nits. PTAL 😊

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (divya-mohan0209) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:27]: I think it would be a comma, not a period @jihoon-seo

Comment (jihoon-seo) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:27]: How about a colon? 😊

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (saschagrunert) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:23]: Let's explain in a sentence why we need to set `COSIGN_EXPERIMENTAL=1` and that it may change in the future.

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (saschagrunert) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:25]: Let's add another section to mention cosigned: https://github.com/sigstore/cosign/tree/main/config

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (PushkarJ) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:26]: added

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (PushkarJ) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md]: added

**Comment by PushkarJ** [comment]
@saschagrunert @jihoon-seo updated as per your feedback

**Review by saschagrunert — APPROVED** [review]
/lgtm

Thanks for working on this @PushkarJ!

**Comment by PushkarJ** [comment]
/assign @bradtopol

(based on bot comment)

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (bradtopol) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:11]: Page looks very good!

Comment (bradtopol) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:11]: Overall this looks very good.  I have not seen the examples folder used before the way you used it. @sftim Is it ok how @PushkarJ stored his examples?

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:42]: The list of signed images should be a release artifact and not stored as an example configuration.

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:31]:
```suggestion
### Verifying images for all control plane components
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:28]:
```suggestion
[Keyless Signatures](https://github.com/sigstore/cosign/blob/main/KEYLESS.md#keyless-signatures).
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:13]:
```suggestion
## {{% heading "prerequisites" %}}

These instructions are for Kubernetes {{< skew currentVersion >}}. If you want to check the
integrity of components for a different version of Kubernetes, check the documentation for
that Kubernetes release.

You will need to have the following tools installed:
- `cosign` ([install guide](https://docs.sigstore.dev/cosign/installation/))
- Go compiler ([install guide](https://go.dev/doc/install)
- `curl` (often provided by your operating system)
```

File: content/en/releases/download.md
Comment (sftim) [inline content/en/releases/download.md:6]: We don't start pages with a second level heading. We shouldn't add this.

File: content/en/releases/download.md
Comment (sftim) [inline content/en/releases/download.md:75]: Can we move the signature information into the Markdown table, perhaps?

File: content/en/releases/download.md
Comment (sftim) [inline content/en/releases/download.md:78]: I'm not clear on what we're saying the reader can do if they follow that link.

Do we mean one of these things:

> You can configure your Kubernetes cluster to verify the integrity of container images, including the
> integrity of Kubernetes components such as the kube-controller-manager. Read [PAGE TITLE]
> to learn how.

or

> You can manually verify the integrity of container images that are part of Kubernetes itself,
> and then [DO THINGS] to ensure that the cluster runs those verified images rather than any
> other.

? If not, what do we mean?

File: content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go
Comment (sftim) [inline content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go:44]: This doesn't seem like the right home for some Golang code.

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (sftim) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:16]:
```suggestion
For a complete list of images that are signed please refer to [Releases](/releases/download/).
```

**Review by sftim — CHANGES_REQUESTED** [review]
Thanks. My big question is: what does the new task page help people to do?

Is it about verifying non-Kubernetes images, given that there is already a cluster, or is it about somehow ensuring that the Kubernetes images that the cluster relies upon are authentic?

Once we know that intent, we can review to see if the new page explains how to achieve this.

File: content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go
Comment (PushkarJ) [reply content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go:44]: Thanks @sftim I was definitely hesitant adding it here but thought that we can always discuss it as part of the PR. @saschagrunert any thoughts on possible locations that could host this? 

Maybe https://github.com/kubernetes-sigs/promo-tools or https://github.com/kubernetes-sigs/release-sdk?

File: content/en/releases/download.md
Comment (PushkarJ) [reply content/en/releases/download.md:78]: This is a good call out for clarity. Let me rewrite the sentence

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (PushkarJ) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:42]: @sftim Maybe this could go in here: https://github.com/kubernetes/website/tree/main/static under a new "rel-eng" directory? @saschagrunert what do you think?

Comment (sftim) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:42]: `rel-eng` is a team detail; we would not expose that to end users.

I'd like to keep SIG Docs away from signing and verification, leaving that as something that SIG Release look after. The places where those artifacts and signatures and lists of signatures belong should change less frequently than our website, and there should be more process attached to updating those.

The way I see it, the _Kubernetes website_ is not where we host artifacts, signatures or lists of signatures. If someone wants to verify a signed artifact, all of that process - bar the page that teaches them how - ought to be hosted separately from the website.

File: content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go
Comment (saschagrunert) [reply content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go:44]: Hm, we already verify that images are signed during the whole release process, which is something we should mention that in the document here.

I can imagine that the code lives in k/release and we make it a dedicated `krel` subcommand. Thoughts on that @puerco?

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (reylejano) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:42]: Is kubernetes.dev a better fit for this page?

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (PushkarJ) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md]: This is being discussed in other comment threads. So resolving this and keeping others open for clarity on current state

File: content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go
Comment (puerco) [reply content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go:44]: I think we can build it into krel. The data to do the verification is already built into the SBOM so we don't have to compile the image list.

Not, that said, if we are going to put out instructions to use krel to verify, we should do two things:

1) Build the subcommand
2) Roll out a plan to release `krel` binaries so that people can fetch them to follow the instructions

If we are OK on this approach I am happy to open the issues to track it (and work on it! )

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:26]:
```suggestion
using the `cosign verify` command:
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:32]: I'd  suggest we use the `note` shortcode for consistency with other pages, i.e,

```
{{< note >}}
something
{{< /note >}}
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (tengqm) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:49]: To move this PR ahead, maybe we can host the text file under `content/en/docs/tasks/administer-cluster/`.
In the long run, this file should be generated and published at https://github.com/kubernetes/kubernetes/releases or somewhere.

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:55]:
```suggestion
the [cosigned](https://docs.sigstore.dev/cosign/kubernetes/#cosigned-admission-controller)
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (tengqm) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:56]:
```suggestion
admission controller. To get started with `cosigned` here are a few helpful resources:
```

File: content/en/releases/download.md
Comment (tengqm) [reply content/en/releases/download.md:75]: Given that the collection of core components is pretty stable at this stage, compiling this into a table seems doable.

File: content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go
Comment (PushkarJ) [reply content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go:44]: @puerco Yay, glad to hear we do not have to maintain the yaml and image list generator. 

But, adding `krel` as a pre-requisite seems like an overhead for users to verify the images. Could we do something like this:
- `krel` subcommand for every version, generates a list of signed images during release process
- That list gets pushed into a GCS bucket or git repo like `k/release` during release of future k8s version where there is separate list for each release
- We `curl` the list from GCS bucket or git repo and provide it to `cosign` in a for loop to verify (similar to current script in this PR)

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (PushkarJ) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:32]: @tengqm can you share an example where this shortcode is used? My understanding (which could be wrong) is that the alert shortcode was the convention

File: content/en/releases/download.md
Comment (PushkarJ) [reply content/en/releases/download.md]: We are using SBOM to get list of images now. So resolving this comment

File: content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go
Comment (PushkarJ) [reply content/en/examples/admin/signed-images/verify-all-signed-control-plane-images.go]: Updated to use SBoM as source of truth for signed images

**Comment by PushkarJ** [comment]
PR is ready for review now, after moving the fetching of signed images to SBoM

https://deploy-preview-32184--kubernetes-io-vnext-staging.netlify.app/releases/download/
https://deploy-preview-32184--kubernetes-io-vnext-staging.netlify.app/docs/tasks/administer-cluster/verify-signed-images/

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (chrisnegus) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:19]:
```suggestion
- Go compiler ([install guide](https://go.dev/doc/install))
```

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (chrisnegus) [inline content/en/docs/tasks/administer-cluster/verify-signed-images.md:54]:
```suggestion
be verified at deploy time using 
```

**Review by chrisnegus — COMMENTED** [review]
Just a couple of small comments. Otherwise, lgtm.

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (tengqm) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:32]: For example, check https://raw.githubusercontent.com/kubernetes/website/main/content/en/docs/concepts/containers/images.md

where you can find:

```
{{< note >}}
This is the recommended approach to run containers based on images
in private registries.
{{< /note >}}
```

Comment (tengqm) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:32]: Also please check the website style guide on this: https://kubernetes.io/docs/contribute/style/style-guide/#shortcodes

File: content/en/releases/download.md
Comment (tengqm) [inline content/en/releases/download.md:73]: Can we remove this line now?

File: content/en/docs/tasks/administer-cluster/verify-signed-images.md
Comment (PushkarJ) [reply content/en/docs/tasks/administer-cluster/verify-signed-images.md:32]: Thank you @tengqm for teaching me something new today!! Updating now.

**Comment by PushkarJ** [comment]
All requested changes are pushed. Also, removed the go compiler as pre-requisite since we are not running any go program anymore. 

Please let me know if I missed anything.

Note: We have not had a beta v1.24 release with all signed images yet! So we may want to keep this ready for merge (with `/approve` and `/lgtm` labels) but `/hold` it until we can verify for real, that the script works with signed release images.

**Comment by sftim** [comment]
/hold
based on https://github.com/kubernetes/website/pull/32184#issuecomment-1090887981

**Review by sftim — COMMENTED** [review]
For components that aren't available as container images, do we provide an official way to verify those yet?
If not, I recommend mentioning that no official mechanism is available there, so that people don't have to read source code etc to answer that question.

**Comment by PushkarJ** [comment]
> For components that aren't available as container images, do we provide an official way to verify those yet? If not, I recommend mentioning that no official mechanism is available there, so that people don't have to read source code etc to answer that question.

I think the [MVP](https://github.com/kubernetes/release/issues/2383) issue documents the scope is restricted to images only. Will let @saschagrunert chime in on if this would be useful to document as current limitation.

@sftim I have also addressed all the rest of the comments.

**Comment by puerco** [comment]
@sftim:  @PushkarJ is right, it's probable we will not get the other artifacts signed for 1.24. It makes sense to leave out the other verification types for now.

**Comment by sftim** [comment]
> leave out the other verification types for now

From a docs point of view, it's useful to let people know what things aren't covered. When we explain that, we try to [avoid statements about the future](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future).

Phrasing like
```
For Kubernetes v{{< skew currentVersion >}}, the only kind of code artifact that you can verify integrity for is a container image, using the experimental signing support.
```
for example (using `v{{< skew currentVersion >}}` so that if eg nothing changes for the v1.25 release, the docs remain relevant).

**Comment by PushkarJ** [comment]
Updated the download.md file with the text suggested by @sftim

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by saschagrunert** [comment]
@sftim can we lift the hold?

**Comment by sftim** [comment]
@PushkarJ did you try this out with the beta release? How did it go?

File: content/en/releases/download.md
Comment (sftim) [inline content/en/releases/download.md:6]: We shouldn't have merged this - level 1 headings aren't allowed in our Markdown.

Comment (saschagrunert) [reply content/en/releases/download.md:6]: Follow-up in https://github.com/kubernetes/website/pull/32895

**Comment by PushkarJ** [comment]
> @PushkarJ did you try this out with the beta release? How did it go?

I did not get a chance to test it before this was unheld and merged, since rc.0 is going to be first release with signed images ETA April 19 :( Either way seems unlikely we may need major overhaul of the script or guide.

**Comment by PushkarJ** [comment]
@sftim just confirmed that the shell script works, but I think adding `| jq` is giving better looking results. Is it still possible to update this page? Demo: https://asciinema.org/a/488826

**Pull request merged**
